src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPEditorImprovementsTypedHandler.java

```diff
@@ -0,0 +1,208 @@
+            document.insertString(offset, String.valueOf(quote));
+            editor.getCaretModel().moveToOffset(offset + 1);
+            return true;
+        }
+
+        return false;
+    }
+
+    private static boolean isStatementTerminatorCharacter(@NotNull PsiFile file, char charTyped) {
+        return LanguageServiceAccessor.getInstance(file.getProject()).hasAny(
+                file.getVirtualFile(),
+                ls -> ls.getClientFeatures().getStatementTerminatorCharacters(file).indexOf(charTyped) > -1
+        );
+    }
+
+    private boolean handleStatementTerminator(@NotNull PsiFile file,
+                                              @NotNull Editor editor) {
+        Document document = editor.getDocument();
+        CaretModel caretModel = editor.getCaretModel();
+        int offset = caretModel.getOffset();
+        if (offset == document.getTextLength()) {
+            return false;
+        }
+
+        CharSequence documentChars = document.getCharsSequence();
+        char nextCharacter = documentChars.charAt(offset);
+        if (isStatementTerminatorCharacter(file, nextCharacter) &&
+            !inStringLiteral(file, documentChars, offset) &&
+            !inComment(file, documentChars, offset)) {
+            caretModel.moveToOffset(offset + 1);
+            return true;
+        }
+
+        return false;
+    }
+
+    @Nullable
+    private static String getStringLiteral(@NotNull PsiFile file,
+                                           @NotNull CharSequence fileChars,
+                                           int offset) {
+        Set<Character> quoteCharacters = LSPIJEditorUtils.getQuoteCharacters(file);
+        if (!ContainerUtil.isEmpty(quoteCharacters)) {
+            // This pattern is basically from any quote character to the same character with no unescaped same quote
+            // character or line break in between
+            Pattern stringLiteralPattern = Pattern.compile("([" + StringUtil.join(quoteCharacters, "") + "])([^$1\\r\\n]|\\\\\\.)*\\1");
+            Matcher stringLiteralMatcher = stringLiteralPattern.matcher(fileChars);
+            while (stringLiteralMatcher.find()) {
+                int stringLiteralStart = stringLiteralMatcher.start();
+                int stringLiteralEnd = stringLiteralMatcher.end();
+                if ((stringLiteralStart < offset) && (offset < stringLiteralEnd)) {
+                    return fileChars.subSequence(stringLiteralStart, stringLiteralEnd).toString();
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private static boolean inStringLiteral(@NotNull PsiFile file,
+                                           @NotNull CharSequence documentChars,
+                                           int offset) {
+        return getStringLiteral(file, documentChars, offset) != null;
+    }
+
+    private static boolean inComment(@NotNull PsiFile file,
```

Comment:
And this provides a quite efficient way to determine if the caret is inside of a line or block comment given the optional line comment prefix and/or block comment prefix/suffix for the language.